### ISSUE TITLE
Duplicate recursion to package file crashes Nuitka

### ISSUE THREAD

**Issue by brupelo** [Nuitka/Nuitka#151]
@kayhayen As I've already commented to you at #139 , here's a little [mcve](https://www.dropbox.com/s/jaxt6dpcqjnbj03/bug_package.zip?dl=1) that will expose you the bug introduced by [https://github.com/Nuitka/Nuitka/commit/f403ee59ca2e5975d517064759b40dc72e9fb472](https://github.com/Nuitka/Nuitka/commit/f403ee59ca2e5975d517064759b40dc72e9fb472).

Here's a little example of execution:

    (py364_32) D:\sources\personal\python\apps\console\pytools> ptime nuitka --output-dir=d:\releases\py364_32_packages\build --no-pyi-file --include-package=bug_package.core --module --jobs=4 D:\sources\personal\sublimetext3\Data\Packages\bug_package\bug_package\core
    Problem with statement at D:\sources\personal\sublimetext3\Data\Packages\bug_package\bug_package\core\a.py:1:
    -> from . import b

    Nuitka:INFO:Interrupted while working on '<Node 'COMPILED_PYTHON_MODULE' with {'filename': 'D:\\sources\\personal\\sublimetext3\\Data\\Packages\\bug_package\\bug_package\\core\\a.py', 'package': 'bug_package.core', 'name': 'a'}>'.
    Traceback (most recent call last):
      File "d:\sources\others\sources_compilers\nuitka\nuitka\__main__.py", line 180, in <module>
        main()
      File "d:\sources\others\sources_compilers\nuitka\nuitka\__main__.py", line 174, in main
        MainControl.main()
      File "d:\sources\others\sources_compilers\nuitka\nuitka\MainControl.py", line 795, in main
        filename = filename
      File "d:\sources\others\sources_compilers\nuitka\nuitka\MainControl.py", line 152, in createNodeTree
        Optimization.optimize(main_module.getOutputFilename())
      File "d:\sources\others\sources_compilers\nuitka\nuitka\optimizations\Optimization.py", line 533, in optimize
        makeOptimizationPass(initial_pass = True)
      File "d:\sources\others\sources_compilers\nuitka\nuitka\optimizations\Optimization.py", line 446, in makeOptimizationPass
        changed = optimizeModule(current_module)
      File "d:\sources\others\sources_compilers\nuitka\nuitka\optimizations\Optimization.py", line 173, in optimizeModule
        changed = optimizeCompiledPythonModule(module)
      File "d:\sources\others\sources_compilers\nuitka\nuitka\optimizations\Optimization.py", line 97, in optimizeCompiledPythonModule
        module.computeModule()
      File "d:\sources\others\sources_compilers\nuitka\nuitka\nodes\ModuleNodes.py", line 467, in computeModule
        trace_collection = self.trace_collection
      File "d:\sources\others\sources_compilers\nuitka\nuitka\nodes\StatementNodes.py", line 165, in computeStatementsSequence
        trace_collection
      File "d:\sources\others\sources_compilers\nuitka\nuitka\nodes\FrameNodes.py", line 195, in computeStatementsSequence

        statement = statement
      File "d:\sources\others\sources_compilers\nuitka\nuitka\optimizations\TraceCollections.py", line 575, in onStatement

        statement.computeStatement(self)
      File "d:\sources\others\sources_compilers\nuitka\nuitka\nodes\AssignNodes.py", line 297, in computeStatement
        trace_collection.onExpression(self.getAssignSource())
      File "d:\sources\others\sources_compilers\nuitka\nuitka\optimizations\TraceCollections.py", line 550, in onExpression
        trace_collection = self
      File "d:\sources\others\sources_compilers\nuitka\nuitka\nodes\ExpressionBases.py", line 1076, in computeExpressionRaw
        expression = sub_expression
      File "d:\sources\others\sources_compilers\nuitka\nuitka\optimizations\TraceCollections.py", line 550, in onExpression
        trace_collection = self
      File "d:\sources\others\sources_compilers\nuitka\nuitka\nodes\ExpressionBases.py", line 1020, in computeExpressionRaw
        trace_collection = trace_collection
      File "d:\sources\others\sources_compilers\nuitka\nuitka\nodes\ImportNodes.py", line 401, in computeExpression
        module_name      = imported_module_name
      File "d:\sources\others\sources_compilers\nuitka\nuitka\nodes\ImportNodes.py", line 291, in _attemptRecursion
        module_package   = module_package
      File "d:\sources\others\sources_compilers\nuitka\nuitka\nodes\ImportNodes.py", line 223, in _consider
        reason          = reason
      File "d:\sources\others\sources_compilers\nuitka\nuitka\importing\Recursion.py", line 169, in recurseTo
        reason          = reason
      File "d:\sources\others\sources_compilers\nuitka\nuitka\importing\Recursion.py", line 68, in _recurseTo
        is_shlib = module_kind == "shlib"
      File "d:\sources\others\sources_compilers\nuitka\nuitka\tree\Building.py", line 1020, in decideModuleTree
        source_ref   = source_ref
      File "d:\sources\others\sources_compilers\nuitka\nuitka\nodes\ModuleNodes.py", line 550, in __init__
        source_ref   = source_ref
      File "d:\sources\others\sources_compilers\nuitka\nuitka\nodes\ModuleNodes.py", line 248, in __init__
        setLocalsDictType(self.module_dict_name, "module_dict")
      File "d:\sources\others\sources_compilers\nuitka\nuitka\nodes\LocalsScopes.py", line 30, in setLocalsDictType
        assert locals_dict_name not in locals_dict_handles, locals_dict_name
    AssertionError: globals_bug_package$core

    Execution time: 1.200 s

Make sure you adjust the paths accordingly before running nuitka. I guess it's totally unrelated but on my local box I've got a pth file living in site-packages where I've got the path poiting out to that package, ie:

    D:\sources\personal\sublimetext3\Data\Packages\bug_package

If you can't reproduce locally just let me know and we'll find better ways.

Thx in advance,
B

**Comment by kayhayen** [comment]
Works as a reproducer indeed.

**Comment by kayhayen** [comment]
This code is in part guilty:

 ```
   if ImportCache.isImportedModuleByPath(module_relpath):
        module = ImportCache.getImportedModuleByPath(module_relpath)

        if module.getCompileTimeFilename().endswith("__init__.py") and \
           module.getPackage() != module_package:
            module = None
    else:
        module = None
```

For what I recall, it is necessary to allow using the same file twice for `__main__` can be imported again, but why the package mismatch would be allowed to occur, I cannot remember at all. Right now I am unconvinced for this to be possible. I guess it has all to do this the command line:

This one crashes: 
> ../bin/nuitka --run --module bug_package/core --verbose --include-module=bug_package.core 2>&1

This one does not: 
> ../bin/nuitka --run --module bug_package/core --verbose --include-module=core 2>&1

So, what is going on is that your command line requires to create module "core" which then contains "subpackage.core" as well, once loaded. That might even work when you place it inside a bug_package module later on. Obviously the use case is not an official one though, and no test of mine covers it.

The search for the module is terminated on first find. When it comes to this spot again, it will import things not only a second time, but a third time. And it will even do a fourth time. If more modules it is as a parent. So the search needs to of course continue past the first find, and ignore it only if it has wrong package_name, but consider later entries.

I am putting a potential fix through the internal CI. Due to release ongoing, factory will get this only after develop is fully built.

Yours,
Kay

**Comment by brupelo** [comment]
> I am putting a potential fix through the internal CI. Due to release ongoing, factory will get this only after develop is fully built.

As a git newbie I'm a little bit curious here, right now I've got `http://git.nuitka.net/Nuitka.git` and `https://github.com/kayhayen/Nuitka/` as remote urls. Those are basically the ones I use to sync my fork, check your commits and play around... So when you say "factory will get this only after develop is fully built", what do you mean exactly? Do you have a CI box over there that's creating the whole bunch of nuitka installers for the general public and once that's done both remote repos are getting in rsynced with your local box?

Mmmm, does that mean you don't need to push your local work to any remote or...? 

Sorry if my question is too silly but because you're handling quite a lot of platforms and the remotes getting updated quite often it seems you've built a nice infrastructure over there so I'm always curious about good practices from other people :)

Anyway, once I'm able to fetch your potential fix I'll continue testing things around, at this moment I'll continue using the good old 70cab3b1 ;D . Usually I use `develop` for testing, unless I've reported some issue and you've pushed to factory.

\o

**Comment by brupelo** [comment]
Today I've seen some new commits on factory and I've tested 04a5424e . It seems the bug is now gone in both the mcve I've provided you and in one of the original packages I was having troubles with.

So yeah, looks fine now, good job!

**Comment by kayhayen** [comment]
This is going to be in the next hotfix release coming soon.

**Comment by kayhayen** [comment]
As for your question, I published the last one in TWN, and this might be good material too. 

Indeed I have a private repo, where I push and private CI picks up. Based on Buildbot, I run many more compilations, basically around the clock, to find regressions from new optimization or codegen changes, and well UI changes too.

Public CI is not aimed at allowing this many compilations. It will be a while before public cloud infrastructure will be donated to Nuitka, although I see it happening some time in the future. This leaves developers with the burden to run tests on their own hardware, and never enough.

My scope is running the CPython test suites on Windows and Linux. These are the 26, 27, 32, 33, 34, 35, 36, 37 suites, and also to get more errors covered, with mismatching Python versions. Often running the 36 tests with 37 and vice versa will extend the coverage. On Windows I compile with and without debug mode, x86 and x64, and it's kind of getting too much. For Linux I have 2 laptops in use, and an ARM CuBox bought from donations, there it's working better, esp. due to ccache being used everywhere.

For memory usage I compile mercurial and observe the memory it used in addition to comparing the mercurial tests to expected outputs its test suite gives. Its a sad day when Mercurial tests find changes in behavior, and luckily that has been rare.

Caching the CPython outputs of tests to compare against is something I am going to make operational these days, maybe after the next hotfix, trying to make things ever faster, but for the time being, ccache.exe and clcache.exe seem to have done wonders.

Anyway, I often tend to let some commits settle in my private repo until they become trusted. Other times I will make bigger changes directly out to factory, because it will be hard to split up the changes later.

I am more conservative with factory right after telling people to try something these. Often I break it on purpose, trying out something. I really consider it a private branch for interacting with me or public CI. I do not recommend to use it.

Then on occasions I am making a sorting of commits on factory and split it into some things that become hotfixes, some things that become current pre-release, and other things that will remain in that proving ground. That is why I typically make hotfix and pre-release at the same times. The git flow suggests doing that and its easy. I do however normally not take things as hotfixes that are on develop already, I hate the duplication of commits.

For me the git bisect is very important. My private commit history is basically a mess and worthless, but on factory I am making very nice organized commits that I will frequently amend, even for the random PyLint cleanup. This allows me when e.g. one test suddenly says "segfault" on Windows to easily find the change that triggers it, look at C code difference, and spot the bug introduced, then amend the commit and be done with it.

My goal is to always have a workable state which is supposed to pass all tests. Obviously I cannot prove it for every commit, but when I know it to not be the case, I tend to make rebased. I even did that for pre-releases gone badly, if it's a problem that affects much code.

As with recent changes, I some times make changes with the "isExperimental()" marker, activating breaking changes only gradually. The C bool type code generation has been there for months in a barely useful form, until it become more polished, and always guarded with a switch, until one day I changed it, and made the necessary fixes retroactively before that switch. 

Then I will remove the experimental code. I feel it's very important and even ideal to be able to always compare to a fully working solution. I am willing to postpone some cleanups until later date as a price, but when something tells me again "This cannot possibly have ever worked"... a command line flag away, I have the answer to compare, plus, extra changes happened in the mean time, they don't add noise to diff outputs of generated C code for example.

Even better, if I decide to make a cleanup as part of making a change more viable to execute, then I get to execute it on stable grounds, covered by the full test suite.

Obviously this takes a lot of hardware to be able to compile this much code on a regular basis. And I really wish I could add one of the new thread rippers to the mix. Anybody donating one to me?

**Comment by brupelo** [comment]
@kayhayen Kay, thank you very much for your detailed answer, it was a really interesting worth reading and it definitely helped me to understand what's going on behind the curtains, just some brief notes about some stuff.

> Public CI is not aimed at allowing this many compilations. It will be a while before public cloud infrastructure will be donated to Nuitka, although I see it happening some time in the future. This leaves developers with the burden to run tests on their own hardware, and never enough.

Disclaimer: Probably other guys will say totally the opossite what I'm going to speak here, even so, here we go:

I've been there... scaling up is a tricky thing on the enterprise world. I've got a strong opinnion about this. I don't think the strategy many companies use of just "throwing out more money to solve the problem" is the right way to go in the long term. Maybe that strategy is alright with huge companies where the benefits outperforms the dev investment... although in many big companies at certain point the control of the product has become less and less obvious and development has become quite an expesive thing, that's really bad. 

But if you're a little company and you've got perfect control about the product itself I'd say investing upfront as much time as possible optimizing your product (testing, building, deploying, ...) before scaling up is really smart. Otherwise you'll end up wasting a lot of cash unnecesarily in the long term.

In fact, in my opinnion there is an "equilibrium point" where the dev effort is no longer profitable and at that time, just at that particular time is the right moment where you can safely "throw out more money to the problem" to scale up.

But again, other guys will say totally the opossite of what I'm saying... if that's so, it'd be interesting to see the maths behind those claims :D

--

One question, I've heard from you many times mentioning this "TWN" thing, what is that? I guess it's neither `https://www.twn.my/` nor `https://www.theweathernetwork.com`, right? :D

Anyway, thank you very much for your post, it's been a really nice reading ;)

**Comment by kayhayen** [comment]
I think one of your main misunderstanding is that this is not "Nuitka Inc." at all, @brupelo but instead a community effort. So comparing to enterprise, talking of customers, leaves me confused.

As for TWN, this is it: http://nuitka.net/categories/NTW.html for which I seem to have changed the name at some point. :)

I committed to more communication and regular one, this way and on Twitter, and so far it's a huge success. I am going in waves with this. But right now I am convinced that excitement about Nuitka is more and more justified and will get more involvement by highly skilled people.

**Comment by brupelo** [comment]
> I think one of your main misunderstanding is that this is not "Nuitka Inc." at all, @brupelo but instead a community effort. So comparing to enterprise, talking of customers, leaves me confused.

Kay, ahahahah :D , fair enough mate... yeah, sure sure :) . But let me tell you, this awesome project is much cooler and much more advanced that many enterprise crap I've seen so far, so... that's why I've get in the mood to talk about these kind of topics. I mean, is it alright to consider these topics to improve mediocre software out there and not really awesome stuff like this? That also left me confused as well... so ;)

> I committed to more communication and regular one, this way and on Twitter, and so far it's a huge success. I am going in waves with this. But right now I am convinced that excitement about Nuitka is more and more justified and will get more involvement by highly skilled people.

That's really awesome Kay, I'm not a "social network" kind of guy myself but it'd be extremely cool if high skilled people will start contributing here. The hardest part of the project is gone thanks to you, now it's mainly the polishing part... So yup, crossing fingers more people will join to the community \o/

TWN, NTW... haha, ok... thanks, cristal clear! Checking now ;)

**Comment by brupelo** [comment]
Ps. Just for the record, customers, end-users, casual-users, all the same thing... At the end of the day this kind of people are the ones that make viral products, give traction to projects, increase popularity of stuff, boost up SEO... all the same thing. If one project is suitable to them, the project will succeed/improve, if not it will progress slowly... 

Let me put it differently, maybe I'm just talking nonsense here but IMHO I think it's very difficult to find high-skilled available people to join, mostly of these people are already running their own projects/stuff and have very little time to get involved with other things... but casual users? That's easier and more achievable.

But tell me as I'm curious, what would you prefer.... 50 casual users testing, opening issues, reporting bugs, giving ideas, doing lightweight stuff or 1 skilled hardcoder helping and making good PRs?

I'm asking you cos if sometime I decide to open-source some of my stuff I'd be wondering what'd be better, making my stuff easier for developers or making it friendlier for casual users.... what's the best choice?

Anyway, hehe... Btw, I've seen before you've marked some of your posts as offtopic, I liked that but i don't find the button here in github to mark my posts as offtopic, mmm :/

**Comment by kayhayen** [comment]
I am not an expert on how to successfully execute an OS project.

As for the marking, it's probably only for the owner of the repo, or now the Nuitka organization members, and it's useful. Allows to retain what was said, but to also allow for a clear view on the issues.

As for preference, I prefer people that are fun to work with. Skill doesn't matter as much to me. I am however convinced that skilled people are often more fun to work with.

**Comment by brupelo** [comment]
I do really love the "marking as offtopic" feature and you're right, probably only the owner of the repo can do it as I didn't find any button over here... so please, feel free to mark my comments as offtopic when they are truly offtopic, that way the issue will be much cleaner for outsider to participate. Sometimes I really feel guilty as abusing github issues to go offtopic is quite convenient actually :/

> As for preference, I prefer people that are fun to work with. Skill doesn't matter as much to me. I am however convinced that skilled people are often more fun to work with.

Yeah, fun factor is definitely one of the main drivers to attract people to contribute on projects. 

As I told you once I don't consider myself a very good team player but this is one skill I'm willing to improve eventually... I guess for me what works better is when I work with honest nice people, if that's the case, I agree with you and I also don't really care about how much "skilled" they are... after all, what exactly is a skill? It's just something that anybody can adquire with time... for instance, I'm really bad at github and a bad team player, so what? those are skills I can perfectly improve with time.

> I am however convinced that skilled people are often more fun to work with.

Mmmm, hehe, not sure about this one... :) . I've worked with really talented people in the past and some of them were like me, really bad team players... But on the other hand of the spectrum, also worked with less "geaky" people and used to have lot of fun working with them as they used to be really friendly... But again, not sure what you mean about "fun" in this context.

Anyway... let's have some fun here :)
